<a href="https://colab.research.google.com/github/nanhoyeon/class2023Fall/blob/main/dimensionality_reduction(PCA).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://medium.com/analytics-vidhya/dimensionality-reduction-principal-component-analysis-d1402b58feb1

## Loading a dataset

In [1]:
import pandas as pd
iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data', names=['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'label'])
iris

sepal_length  sepal_width  petal_length  petal_width           label
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

In [2]:
X = iris.drop(['label'], axis=1)
X

sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

## Principal Component Analysis (PCA)

In [3]:
from sklearn.decomposition import PCA
model = PCA(n_components=2)  # n_components specifies the new dimensionality
X_reduced = model.fit(X).transform(X) # or just pca.fit_transform(X)
pd.DataFrame(X_reduced)

0         1
0   -2.684207  0.326607
1   -2.715391 -0.169557
2   -2.889820 -0.137346
3   -2.746437 -0.311124
4   -2.728593  0.333925
..        ...       ...
145  1.944017  0.187415
146  1.525664 -0.375021
147  1.764046  0.078519
148  1.901629  0.115877
149  1.389666 -0.282887

[150 rows x 2 columns]

PCA(n_components=2)
PCA 2개만 쓰겠다는 말
5로 하면 error가 나올거고, 여기서 최대는 4

원점을 데이터 한 가운데로 끌어당겼을 때의 축 기준으로
긴 데이터에 해당하는 축이 PC1, 좁은 폭이 PC2

In [4]:
print('Components:', model.components_)
print('Explained Variance:', model.explained_variance_)
print('Explained Variance Ratio:', model.explained_variance_ratio_ )

Components: [[ 0.36158968 -0.08226889  0.85657211  0.35884393]
 [ 0.65653988  0.72971237 -0.1757674  -0.07470647]]
Explained Variance: [4.22484077 0.24224357]
Explained Variance Ratio: [0.92461621 0.05301557]


model.components -> 숫자 4개 나오는 건 뭐지?
4차원이기 때문에 방향을 말하기 위해 숫자 4개가 나오는 것이다. []가 2개 있는 것은 각각 PC1과 PC2다.
PC1과 PC2 크기와 비율이 다른데, 무조건 PC1이 크게 나온다.

Explained Variance Ratio: [0.92461621 0.05301557]
여기서 0.924는 원래 숫자 4개로 꽃을 설명하다가 PC1, 즉 숫자 하나 가지고 설명했을 때의 정확도를 의미한다(92.4%) 만약 숫자를 하나 더 추가한다면 0.97... 거의 98% 커버 가능하다.(숫자 4개를 추가할 필요가 없는 이유)

In [5]:
X_approx = model.inverse_transform(X_reduced)
pd.DataFrame(X_approx)

0         1         2         3
0    5.087182  3.513156  1.402043  0.211056
1    4.750155  3.153664  1.462541  0.236932
2    4.708232  3.191519  1.307469  0.171933
3    4.645984  3.052915  1.460831  0.236367
4    5.075937  3.522147  1.362737  0.194581
..        ...       ...       ...       ...
145  6.669315  3.030827  5.390916  1.882264
146  6.148781  2.654828  5.131424  1.774158
147  6.532745  2.966170  5.255898  1.825818
148  6.607020  2.982112  5.367182  1.872398
149  6.160096  2.733248  4.998738  1.718473

[150 rows x 4 columns]

이거는 숫자 2개 가지고 다시 4개로 복귀?하는

In [6]:
X

sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]